src/main.c

```diff
@@ -4147,18 +4254,31 @@ H2O_NORETURN static void *run_loop(void *_thread_index)
     /* setup listeners */
     for (i = 0; i != conf.num_listeners; ++i) {
         struct listener_config_t *listener_config = conf.listeners[i];
-        int fd = listener_config->fds.entries[thread_index];
         listeners[i] = (struct listener_ctx_t){i,
                                                {&conf.threads[thread_index].ctx, listener_config->hosts, NULL, NULL,
                                                 listener_config->proxy_protocol, &conf.threads[thread_index].memcached}};
-        if (listener_config->ssl.size != 0) {
+        if (listener_config->ssl.size != 0 && !is_reverse_listener(listener_config)) {
             listeners[i].accept_ctx.ssl_ctx = listener_config->ssl.entries[0]->identities[0].ossl;
             listeners[i].accept_ctx.http2_origin_frame = listener_config->ssl.entries[0]->http2_origin_frame;
         }
-        listeners[i].sock = h2o_evloop_socket_create(conf.threads[thread_index].ctx.loop, fd, H2O_SOCKET_FLAG_DONT_READ);
-        listeners[i].sock->data = listeners + i;
+        if (is_reverse_listener(listener_config)) {
```

Comment:
See my comment on reverse_tunnel.h. I do not think we need init and start_listening, or the need to retain the information in `listener_config`.

We can just setup the reverse listener and start listening after the ordinary listeners start accepting.

Reply:
I basically agree but would you mind clarifying what this means?

> start listening after the ordinary listeners start accepting.

Is it a requirement to start listening after other listeners start accepting? If so why?

Reply:
For ordinary listeners, we do the following:
1. setup the per-thread state for accepting connections in all the threads (this might hard fail)
2. setup graceful shutdown logic in thread zero
3. start accepting connections

But in the case of reverse http, there is no need to have separate steps (1 and 3), because the reverse channel for "accepting HTTP connections" is opened and reopened after the server starts running. Failures to setup the reverse channel is expected, and we will be treating that as a softfail, trying to reconnect periodically.

That's the reason everything can be done in step 3, and that we do not need init and start_listening as separate functions is reverse_channel.h.

Reply:
Addressed in https://github.com/h2o/h2o/pull/3405/commits/04307140c0b03eff2983fa9fbb6e2e9f1157388a.

> See https://github.com/h2o/h2o/pull/2913 regarding why we do have these steps.

#2913 and #2915 are a mix of several issues and fixes, and just by glancing them I couldn't find an answer to the question of why the initialization and the start of the accept process need to be separated. Appreciate it if you could explain it to me when you have time.

Reply:
The problem descibed in #2913 is as follows:
* Once we start accepting connections and receive SIGTERM, we need to shutdown gracefully. In order to do so, we broadcast the message using `h2o_multithread_send_message`.
* But we cannot call`h2o_multithread_send_message` before all the threads have completed setting up the receiver.
* Therefore, installation of the signal handler must happen after all the threads have completed their setup phase, but before they start accepting the connections.

Reply:
Thank you for the explanation. And reading your comment I remember that I have to add graceful shutdown handling also for reverse-listening connections.

Reply:
Oh, it looks like I had already marked that as TODO :yum:
https://github.com/h2o/h2o/pull/3405/files#diff-e0cf5b28d9b6b600f0af2bc78e8fd30ec675fd731a5da86f0c4283ffc0e40176R4353